server/src/main/java/eu/fasten/server/plugins/kafka/FastenKafkaPlugin.java

```diff
@@ -264,32 +265,43 @@ public void handleConsuming() {
-                    // However, if local storage is failing we can't store the CG's either and that's already a problem.
-                    logger.error("Trying to store the hash of a record, but failed due to an IOException", e);
-                } finally { // Event if we hit an IOException, we will execute this finally block.
-                    if (consumeTimeoutEnabled) {
-                        consumeWithTimeout(record.value(), consumeTimeout, exitOnTimeout);
-                    } else {
-                        plugin.consume(record.value());
+
+        try {
+            if (localStorage != null) { // If local storage is enabled.
+                if (localStorage.exists(record.value(), record.partition())) { // This plugin already consumed this record before, we will not process it now.
+                    logger.info("Already processed record with hash: " + localStorage.getSHA1(record.value()) + ", skipping it now.");
+                    plugin.setPluginError(new ExistsInLocalStorageException("Record already exists in local storage. Most probably it has been processed before and the pod crashed."));
+                } else {
+                    try {
+                        localStorage.store(record.value(), record.partition());
+                    } catch (IOException e) {
+                        // We couldn't store the message SHA. Will just continue processing, but log the error.
+                        // This strategy might result in the deadlock/retry behavior of the same coordinate.
+                        // However, if local storage is failing we can't store the CG's either and that's already a problem.
+                        logger.error("Trying to store the hash of a record, but failed due to an IOException", e);
+                    } finally { // Event if we hit an IOException, we will execute this finally block.
+                        if (consumeTimeoutEnabled) {
+                            consumeWithTimeout(record.value(), consumeTimeout, exitOnTimeout);
+                        } else {
+                            plugin.consume(record.value());
+                        }
                     }
                 }
+            } else { // If local storage is not enabled.
+                if (consumeTimeoutEnabled) {
+                    consumeWithTimeout(record.value(), consumeTimeout, exitOnTimeout);
+                } else {
+                    plugin.consume(record.value());
+                }
             }
-        } else { // If local storage is not enabled.
-            if (consumeTimeoutEnabled) {
-                consumeWithTimeout(record.value(), consumeTimeout, exitOnTimeout);
-            } else {
-                plugin.consume(record.value());
+        } catch (RuntimeException e) {
+            // In rare circumstances, plug-ins throw UnrecoverableError to crash and therefore K8s will restart the plug-in.
+            if (e instanceof UnrecoverableError) {
```

Comment:
Please do not use this check. Instead have a separate catch clause for `UnrecoverableError` first, and then have a second catch clause for the more general `RuntimeException`. The clauses will be processed in order and only the first match is being executed.

Reply:
Addressed in e116cdb.